modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/metadata/ColocationGroup.java

```diff
@@ -194,30 +186,73 @@ public ColocationGroup colocate(ColocationGroup other) throws ColocationMappingE
         return new ColocationGroup(sourceIds, nodeNames, assignments);
     }
 
+    private List<NodeWithTerm> intersect(List<NodeWithTerm> assign0, List<NodeWithTerm> assign1, @Nullable Set<String> filter, int partId)
+            throws ColocationMappingException {
+        List<NodeWithTerm> intersection = new ArrayList<>();
+
+        for (NodeWithTerm nodeWithTerm : assign0) {
+            if (filter != null && !filter.contains(nodeWithTerm.name())) {
+                continue;
+            }
+
+            for (NodeWithTerm otherNodeWithTerm : assign1) {
+                if (!otherNodeWithTerm.name().equals(nodeWithTerm.name())) {
+                    continue;
+                }
+
+                if (nodeWithTerm.term() != otherNodeWithTerm.term()) {
+                    throw new ColocationMappingException("Raft group primary replica term has been changed during mapping ["
+                            + "part=" + partId
+                            + ", leader=" + nodeWithTerm.name()
+                            + ", expectedTerm=" + nodeWithTerm.term()
+                            + ", actualTerm=" + otherNodeWithTerm.term() + ']');
+                }
+
+                intersection.add(otherNodeWithTerm);
+            }
+        }
+
+        return intersection;
+    }
+
+    private List<NodeWithTerm> filter(List<NodeWithTerm> assignment, Set<String> filter) {
+        List<NodeWithTerm> res = new ArrayList<>();
+
+        if (nullOrEmpty(assignment) || nullOrEmpty(filter)) {
+            return Collections.emptyList();
+        }
+
+        for (NodeWithTerm nodeWithTerm : assignment) {
+            if (!filter.contains(nodeWithTerm.name())) {
+                continue;
+            }
+
+            res.add(nodeWithTerm);
+        }
+
+        return res;
+    }
+
     /**
      * Constructor.
```

Comment:
Let's fix javadoc

Reply:
Done.

Reply:
I mean remove `TODO` and write javadoc - let's try improve our documentation step by step,